pyathena/common.py

```diff
@@ -142,24 +142,53 @@ def _build_start_query_execution_request(self, query, work_group=None, s3_stagin
             })
         return request
 
-    def _execute(self, operation, parameters=None, work_group=None, s3_staging_dir=None):
+    def _find_previous_query_id(self, request, work_group, cache_size):
+        next_token = None
+        while cache_size > 0:
+            n = min(cache_size, 50)  # 50 is max allowed by AWS API
+            cache_size -= n
+            request_params = {'MaxResults': n}
+            if work_group is not None:
+                request_params['WorkGroup'] = work_group
+            if next_token is not None:
+                request_params['NextToken'] = next_token
+            response = self.connection._client.list_query_executions(**request_params)
+            query_ids = response['QueryExecutionIds']
+            next_token = response['NextToken']
+            query_executions = self.connection._client.batch_get_query_execution(
+                QueryExecutionIds=query_ids
+            )['QueryExecutions']
+            for execution in query_executions:
+                queries_match = execution['Query'] == request['QueryString']
+                succeeded = execution['Status']['State'] == 'SUCCEEDED'
+                is_dml = execution['StatementType'] == 'DML'
+                if queries_match and succeeded and is_dml:
+                    return execution['QueryExecutionId']
+
+    def _execute(self, operation, parameters=None, work_group=None, s3_staging_dir=None,
+                 cache_size=0):
         query = self._formatter.format(operation, parameters)
         _logger.debug(query)
 
         request = self._build_start_query_execution_request(query, work_group, s3_staging_dir)
         try:
-            response = retry_api_call(self._connection.client.start_query_execution,
-                                      config=self._retry_config,
-                                      logger=_logger,
-                                      **request)
-        except Exception as e:
-            _logger.exception('Failed to execute query.')
-            raise_from(DatabaseError(*e.args), e)
-        else:
-            return response.get('QueryExecutionId', None)
+            query_id = self._find_previous_query_id(request, work_group, cache_size)
+        except Exception:
+            _logger.warn('Failed to check the cache. Moving on without cache.')
```

Comment:
I think that the process of receiving an exception here is not appropriate.
I think it should be supported by the `_find_previous_query_id` method.